### PR TITLE
IAM USer Policy resource and datsource, Enhanced accounts to export u…

### PR DESCRIPTION
**Pull request by Praveengostu**
…sers in account

### CODE REVIEW

File: ibm/data_source_ibm_account.go [hunk 1]
```diff
@@ -34,6 +58,26 @@ func dataSourceIBMAccountRead(d *schema.ResourceData, meta interface{}) error {
...
+
+	accountv1Client, err := meta.(ClientSession).BluemixAcccountv1API()
+	if err != nil {
+		return err
+	}
+	accountUsers, err := accountv1Client.Accounts().GetAccountUsers(account.GUID)
+	if err != nil {
+		return fmt.Errorf("Error retrieving users in account: %s", err)
+	}
+	accountUsersMap := make([]map[string]interface{}, 0, len(accountUsers))
```
Comment (ashishth09) [inline ibm/data_source_ibm_account.go]: []map[string]string{} should be used

File: ibm/data_source_ibm_account.go [hunk 2]
```diff
@@ -34,6 +58,26 @@ func dataSourceIBMAccountRead(d *schema.ResourceData, meta interface{}) error {
...
+	for _, user := range accountUsers {
+		accountUser := make(map[string]interface{})
+		accountUser["id"] = user.Id
+		accountUser["email"] = user.Email
+		accountUser["state"] = user.State
+		accountUser["role"] = user.Role
+		accountUsersMap = append(accountUsersMap, accountUser)
```
Comment (ashishth09) [inline ibm/data_source_ibm_account.go]: accountUsersMap[i] = accountUser would be cleaner to use I guess once you don't ignore the index in the for loop

File: ibm/data_source_ibm_iam_user_policy.go [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+							Computed: true,
+						},
+						"roles": {
+							Type:     schema.TypeList,
+							Computed: true,
+							Elem: &schema.Resource{
+								Schema: map[string]*schema.Schema{
+									"id": {
+										Type: schema.TypeString,
+
```
Comment (ashishth09) [inline ibm/data_source_ibm_iam_user_policy.go]: Redundant space

File: ibm/data_source_ibm_iam_user_policy.go [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+func dataSourceIBMIAMUserPolicyRead(d *schema.ResourceData, meta interface{}) error {
+	iamClient, err := meta.(ClientSession).IAMAPI()
+	if err != nil {
+		return err
+	}
+
+	scope := d.Get("account_guid").(string)
+	ibmId := d.Get("ibm_id").(string)
+	userId, err := getIBMUniqueIdOfUser(scope, ibmId, meta)
+	log.Println("praveen0", userId)
```
Comment (ashishth09) [inline ibm/data_source_ibm_iam_user_policy.go]: Please remove this log statement.

File: ibm/data_source_ibm_iam_user_policy.go [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+	if userId == "" || err != nil {
+		return fmt.Errorf("User doesnot exist in the account", err)
```
Comment (ashishth09) [inline ibm/data_source_ibm_iam_user_policy.go]: We should the ibmid in the error to make it to more clear to the user

File: ibm/data_source_ibm_iam_user_policy.go [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+	}
+
+	accessPolicyList, err := iamClient.IAMPolicy().List(scope, userId)
+	if err != nil {
+		return fmt.Errorf("Error retrieving list of policies %s", err)
+	}
+
+	policies := accessPolicyList.Policies
+	log.Println("praveen1", len(policies))
+	accountPolicyListMap := make([]map[string]interface{}, 0, len(policies))
```
Comment (ashishth09) [inline ibm/data_source_ibm_iam_user_policy.go]: Make a map of map[string]string

File: ibm/data_source_ibm_iam_user_policy.go [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+		rolesMaps[OPERATOR_ID] = OPERATOR
+		roles := flattenIAMPolicyRoles(policy.Roles, rolesMaps)
+		resources := flattenIAMPolicyResource(policy.Resources, iamClient)
+
+		l := map[string]interface{}{
+			"id":        policy.ID,
+			"roles":     roles,
+			"resources": resources,
+		}
+		log.Println("praveen2", l)
```
Comment (ashishth09) [inline ibm/data_source_ibm_iam_user_policy.go]: Please remove this log statement

File: ibm/resource_ibm_iam_user_policy.go [hunk 8]
```diff
@@ -0,0 +1,376 @@
...
+
+import (
+	"fmt"
+	v1 "github.com/IBM-Bluemix/bluemix-go/api/iampap/iampapv1"
+	"github.com/IBM-Bluemix/bluemix-go/bmxerror"
+	"github.com/hashicorp/terraform/helper/schema"
+	"strings"
+)
+
+const (
```
Comment (ashishth09) [inline ibm/resource_ibm_iam_user_policy.go]: These should be moved to the bluemix-go SDK

File: ibm/structures.go [hunk 9]
```diff
@@ -137,3 +139,64 @@ func flattenServiceInstanceCredentials(keys []mccpv2.ServiceKeyFields) []interfa
...
 	return out
 }
+
+func flattenIAMPolicyResource(list []iampapv1.Resources, iamClient iampapv1.IAMPAPAPI) []map[string]interface{} {
+	result := make([]map[string]interface{}, 0, len(list))
+	for _, i := range list {
+		name := i.ServiceName
+		var serviceName string
+		if name != "" {
+			serviceName, _ = iamClient.IAMService().GetServiceDispalyName(name)
```
Comment (ashishth09) [inline ibm/structures.go]: Are we ignoring the error here?

File: ibm/structures.go [hunk 10]
```diff
@@ -137,3 +139,64 @@ func flattenServiceInstanceCredentials(keys []mccpv2.ServiceKeyFields) []interfa
...
+		} else {
+			serviceName = "All Identity and Access enbled services"
+		}
```
Comment (ashishth09) [inline ibm/structures.go]: If we are assigning `All Identity and Access enbled services`.(Also note that  that spelling is incorrect) then we should define it as default value in the schema

File: ibm/data_source_ibm_iam_user_policy.go [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+						"id": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"roles": {
+							Type:     schema.TypeList,
+							Computed: true,
+							Elem: &schema.Resource{
+								Schema: map[string]*schema.Schema{
+									"id": {
```
Comment (ashishth09) [inline ibm/data_source_ibm_iam_user_policy.go]: `id` should be appropriately called `type` or `name`

File: ibm/resource_ibm_iam_user_policy.go [hunk 12]
```diff
@@ -0,0 +1,376 @@
...
+	policyId := d.Id()
+	iamPolicy, err := iamClient.IAMPolicy().Get(accountGuid, userId, policyId)
+	if err != nil {
+		return fmt.Errorf("Unable to read policy:%s", err)
+	}
+
+	resources := iamPolicy.Resources
+	roles := iamPolicy.Roles
+
+	d.Set("roles", flattenIAMPolicyRoles(roles, rolesMaps))
```
Comment (ashishth09) [inline ibm/resource_ibm_iam_user_policy.go]: I think we don't need to pass rolesMaps in the function. Have that map defined somewhere in var

File: ibm/structures.go [hunk 13]
```diff
@@ -137,3 +139,64 @@ func flattenServiceInstanceCredentials(keys []mccpv2.ServiceKeyFields) []interfa
...
+		region := i.Region
+		resourceType := i.ResourceType
+		resource := i.Resource
+		spaceId := i.SpaceId
+		organizationId := i.OrganizationId
+		serviceInstance := i.ServiceInstance
+		ok, element := contains(serviceName, region, resourceType, resource, spaceId, organizationId, result)
```
Comment (ashishth09) [inline ibm/structures.go]: Do we really need `contains` function?

**Comment by Praveengostu** [comment]
Due to some merging conflicts in vendor.json.. closing this and will raise a new PR

**Pull request closed without merging**